akka-stream/src/main/scala/ResponseStage.scala

```diff
@@ -99,9 +102,11 @@ private[akkastream] class ResponseStage[T, Out](
 
       override def postStop(): Unit = {
         killLast()
+        shutdownPromise.tryCompleteWith(State.materialized)
```

Comment:
To have a really accurate mat val, would do it in the stage, which can complete with a failure if any encountered, or with successful `State`.

Reply:
So, if I understand correctly: `killLast` would return a Try[Unit] (or Future[Unit]) that would be mapped over to complete the promise? (sound good to me)

Reply:
On second thought, maybe the Future should not be failed for the sole reason that resource clean-up failed. In this case, we can do like for source in `akka.stream.scaladsl.FileIO` and return a kind of `akka.stream.IOResult` that incorporate any exception encountered during cleanup, and that way, still have other useful info as well, eg.:
`case class State(count: Long, status: Try[Closed])` (to mimic `IOResult`, `count` would be the total number of document emitted).

Reply:
The return type of the current stage function should not need to be changed, but the `Promise` be completed when the completion state is known from there.

As for the `State`, it would need to make sure there is a coherent count, and that there is no `API` regression/compat issue about this type already part of the API.